In [1]:
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings('ignore', category=DeprecationWarning, module='.*/IPython/.*')
warnings.filterwarnings('ignore', category=DeprecationWarning, module='pyLDAvis')

import pyLDAvis
import pyLDAvis.sklearn

from sklearn.feature_extraction.text import CountVectorizer
from sklearn.decomposition import LatentDirichletAllocation

pyLDAvis.enable_notebook()

In [2]:
import warnings
warnings.filterwarnings("ignore")

In [3]:
import spacy, re
import pandas as pd
import numpy as np
from sqlalchemy import create_engine
from string import punctuation, printable
from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer
from sklearn.decomposition import NMF, LatentDirichletAllocation

In [4]:
engine = create_engine('postgresql://jordanhelen:password@localhost:5432/firewise')
sql = "select description, event_id from old_community_event where description is not null or description <> '';"

In [5]:
def df_from_sql(sql_code_str, engine):
    print("Bringing data in...")
    df = pd.read_sql(sql_code_str, engine)
    contents = df[df.columns[0]].values
    return df, contents

In [6]:
df, contents = df_from_sql(sql, engine)

Bringing data in...


In [16]:
def clean_text(contents):
    print("Lemmatizing, removing stop words, cleaning text...")
    punc_dict = {ord(punc): None for punc in punctuation}
    nlp = spacy.load("en")
    for i, line in enumerate(contents):
        line = line.translate(punc_dict)
        clean_doc = "".join([char for char in line if char in printable])
        line = nlp(clean_doc)
        line_list = [re.sub("\W+","",token.lemma_.lower()) for token in line if token.is_stop == False]
        line_list = [token for token in line_list if token not in ('2015','2014','2016','fire','firewise')]
        contents[i] = ' '.join(line_list)
    return contents

In [17]:
contents = clean_text(contents)

Lemmatizing, removing stop words, cleaning text...


In [38]:
def tf(contents):
    print("Extracting tf features for LDA...")
    tf_vectorizer = CountVectorizer(max_df=.99, min_df=0.01, max_features=150)
    X = tf_vectorizer.fit_transform(contents)
    feature_names = tf_vectorizer.get_feature_names()
    return X, feature_names, tf_vectorizer

In [39]:
X, features_name, tf_vectorizer = tf(contents)

Extracting tf features for LDA...


In [46]:
def run_lda(X):
    print("Running LDA model...")
    model = LatentDirichletAllocation(n_components=7,max_iter=100,learning_method='online',learning_offset=50.,random_state=0)
    return model.fit(X)

In [47]:
model = run_lda(X)

Running LDA model...


In [48]:
pyLDAvis.sklearn.prepare(model, X, tf_vectorizer, R=20)

PreparedData(topic_coordinates=            Freq  cluster  topics         x         y
topic                                                
2      18.554590        1       1 -0.027326 -0.235302
5      17.526591        1       2 -0.201394  0.212210
1      16.270395        1       3  0.208199  0.026264
0      12.695128        1       4  0.171264 -0.156786
3      12.672066        1       5  0.090273  0.193026
4      11.330664        1       6 -0.351686 -0.096523
6      10.950566        1       7  0.110670  0.057111, topic_info=     Category         Freq            Term        Total  loglift  logprob
term                                                                     
114   Default   898.000000        property   898.000000  20.0000  20.0000
63    Default   873.000000            fuel   873.000000  19.0000  19.0000
113   Default   847.000000         project   847.000000  18.0000  18.0000
102   Default   887.000000          people   887.000000  17.0000  17.0000
91    Default  1313.000000         meeting  1313.000000  16.0000  16.0000
19    Default   947.000000           brush   947.000000  15.0000  15.0000
39    Default  2665.000000             day  2665.000000  14.0000  14.0000
141   Default   928.000000            tree   928.000000  13.0000  13.0000
56    Default  1479.000000           event  1479.000000  12.0000  12.0000
16    Default   765.000000           booth   765.000000  11.0000  11.0000
25    Default   715.000000           clean   715.000000  10.0000  10.0000
11    Default   794.000000          attend   794.000000   9.0000   9.0000
23    Default   580.000000         chipper   580.000000   8.0000   8.0000
127   Default   651.000000          safety   651.000000   7.0000   7.0000
146   Default   680.000000        wildfire   680.000000   6.0000   6.0000
22    Default   818.000000            chip   818.000000   5.0000   5.0000
98    Default   500.000000           owner   500.000000   4.0000   4.0000
133   Default   483.000000           space   483.000000   3.0000   3.0000
71    Default  1618.000000            hold  1618.000000   2.0000   2.0000
147   Default   955.000000            work   955.000000   1.0000   1.0000
93     Topic1   559.007497      mitigation   559.874062   1.6829  -3.1359
28     Topic1   281.829503       committee   282.695895   1.6814  -3.8207
135    Topic1   266.505052           state   267.371058   1.6812  -3.8766
123    Topic1   254.045333          review   254.910945   1.6811  -3.9245
15     Topic1   250.735133           board   251.601419   1.6810  -3.9376
124    Topic1   204.247181            risk   205.113048   1.6802  -4.1427
121    Topic1   181.132843  representative   181.998449   1.6797  -4.2628
146    Topic1   675.905785        wildfire   680.115417   1.6782  -2.9460
34     Topic1   167.050895         council   179.237046   1.6140  -4.3437
108    Topic1   166.742268    preparedness   188.692519   1.5608  -4.3456
...       ...          ...             ...          ...      ...      ...
32     Topic6    87.069426         conduct   266.928873   1.0574  -4.5021
72     Topic6   354.381822            home  1193.219223   0.9636  -3.0984
45     Topic6   146.960734         discuss   672.967395   0.6561  -3.9787
30     Topic6   364.768408       community  3402.110642  -0.0552  -3.0696
90     Topic6    89.862213            meet   422.527305   0.6297  -4.4705
133    Topic7   482.188589           space   483.042251   2.2100  -2.7564
42     Topic7   403.498969      defensible   404.352490   2.2097  -2.9345
99     Topic7   327.157956            park   328.012014   2.2092  -3.1443
53     Topic7   308.806136       emergency   309.659571   2.2090  -3.2020
64     Topic7   288.735951              fw   289.589627   2.2088  -3.2692
51     Topic7   252.033107     educational   252.886910   2.2084  -3.4051
131    Topic7   229.782838            sign   230.636618   2.2081  -3.4976
55     Topic7   216.036426      evacuation   216.890122   2.2078  -3.5592
86     Topic7   438.297671           local   601.236906   1.8957  -

In [18]:
#Save HTML Code
p = pyLDAvis.sklearn.prepare(model, X, tf_vectorizer, R=20)
pyLDAvis.save_html(p, 'lda.html')

## LDA Indicies from above graph

Topic 1 [Distribution Event]: community booth information event annual hold meeting sign attend home

Topic 2 [Education Event]: community presentation property resident fuel wildfire forest service provide discuss

Topic 3 [Home assessment]: home homeowner assessment member information wildland conduct answer question risk

Topic 4 [Community Preparedness]: day community program material county event hold mitigation member chip

Topic 5 [Mitigation Event]: brush chip road volunteer year hour community area property chipper

### From LDA Python Script and mapped to counts below
Topic # 0: day community program material county event hold mitigation member chip

Topic # 1: home homeowner assessment member information wildland conduct answer question risk

Topic # 2: community booth information event annual hold meeting sign attend home

Topic # 3: brush chip road volunteer year hour community area property chipper

Topic # 4: community presentation property resident fuel wildfire forest service provide discuss

 topics | count 
 
    0 |  2364
    1 |   560
    2 |  3131
    3 |  1106 
    4 |  2095